### DISCUSSION TITLE
Custom Aggregate Functions in Data Table Component

### DISCUSSION THREAD

**Discussion by nickfanion** [evidence-dev/evidence#2491] [Ideas]
### Description

I think the Data Table component needs the option to specify custom aggregation functions or formulas to display accurate total and subtotal rows.

### What problem would this solve?

Without the ability to specify a custom aggregation function, some aggregations cannot be calculated accurately. For example, take percent change:

`( new_value - old_value ) / old_value`

I have a data table with four columns:

- product_name
- sales
- sales_year_ago
- sales_yoy_percent_chg (derived from sales and sales_year_ago using the percent change calculation in sql)

The aggregate in the total row correctly sums the sales and sales_year_ago columns, but none of the existing aggregation functions correctly display sales_yoy_percent_chg. The sales and sales_year_ago columns need to be aggregated with sum first, and then the calculation above applied. Applying a mean or weighted mean aggregation to the product percent change values does not yield the correct result.

EDIT: I forgot to add that custom aggregation values is a solution for tables with a single total row, but does not work for the unique aggregations necessary in a groupBy subtotal.

**Comment by nickfanion** [comment]
On Slack, Sean pointed out to me that setting `weightCol` to `old_value` instead of `new_value` will aggregate percent change values correctly with `totalAgg` set to `weightedMean`. Custom aggregation functions might still be a useful feature for other use cases, but I haven't run into others that can't be solved by the built-in aggregation functions.

**Reply by hughess** [reply]
Thanks @nickfanion!

We've got an issue open to improve the documentation for this. Thanks for flagging it
